Backup and Restore Documentation Updates (#7)
mhira-project/documentation: docs/data-and-backup.md

# Data and Backup

## Data Storage Location

> The location where mhira-docker installation will store its data is controlled by environment variable `DATA_SAVE_PATH`.

Reviewer: ```suggestion
The location where mhira-docker installation will store its data is controlled by environment variable `DATA_SAVE_PATH` in the .env file.
```

```dotenv
# Default location under current user home directory
DATA_SAVE_PATH=~/.mhira-docker/data
```

> This is where all stateful containers (containers which persist data) will store their configuration and data.

Reviewer: wouldn't you need to use the word 'docker volumes' here?

Author: For mhira case, the container persistance mechanism uses "bind-mounts" rather than "volumes" so it not applicable for this case. The difference expanded [here on stackoverflow](https://stackoverflow.com/questions/34357252/docker-data-volume-vs-mounted-host-directory)

In a typical installation, the folder contents will be as listed below.

![Data-Directory](/resources/images/data-directory.png "Data Directory")

***Important note on MHIRA running user context:***
> As the data directory is a relative path to the current user's home directory, the user under which `MHIRA` is running is of importance.
>
> For example: if `MHIRA` docker-compose is run as root user, the data save path will be under the root user's home directory, i.e: `/root/.mhira-docker/data`.
>

> > Switching running mhira using one user to another can result in apparent data loss, as the new user context will create a new data directory under their home directory, which will not include data from the old data directory.

Reviewer: ```suggestion
> Switching a running instance of mhira to another user can result in apparent data loss, as the new user context will create a new data directory under their home directory, which will not include data from the old data directory.
```

>
> To avoid this, MHIRA system admin can replace the `DATA_SAVE_PATH` with an absolute path pointing to a directory they wish to persist data.

## Backup data

> 

Reviewer: This summary section from  ## Backup data up to ### Automated daily backups
is confusing. I would delete it and get started more pragmatically with the next section directly. This seems more practicle.

> MHIRA data backup uses standard file formats: TAR for creating archives and GZIP for compression. These formats are natively supported on linux, macOS and other operating systems. This makes it possible to easily extract files from the archives.

Reviewer: 0

> The backup utility simply creates a TAR archive of the MHIRA data directory and saves it in configured data backup path.

Reviewer: 0

To restore from backup, you may have multiple archives created at different times (we call them versions). Choose one of them (the latest if you want to restore the most recent copy of your data) and extract it. Copy the extracted data to  the configured MHIRA data save path.

The next subsections give more details on the provided automated backups utility, how to do manual backups, and restore process.

### Automated daily backups

The `MHIRA` docker installation comes with a `backup` helper container, that manages daily backups.

> 

Reviewer: This part needs to be elaborated-- 
Mention what is in this backup container. 
I understand, it is a container with Ubuntu and then cron is installed. 
The documantion should explain how it is working and what tools are being used. . 

The idea with having a backup utility in its own container is good. 
However, your setup is not understanable to people without good coding skills. 
For me, it would take hours to understand how I can modify stuff. 

Why not use something with a GUI? like duplicati (but that one has problems with restoring)

Reviewer: On the other hand, if we want to enable more options, crontab and rsync can be combined. 
Having a backup on the same harddrive is useless

> * The default backup schedule is daily at midnight.
>
> * The backup location is controlled by by `DATA_BACKUP_PATH` key in the `.env` file.
>
> * Default value for `DATA_BACKUP_PATH` is `~/.mhira-docker/backups`

> 

Reviewer: "Warning: You should backup your data to a different hard drive, machine, or even location to make sure your data can be restored in case of hard drive failure or other incidents. "

Is there any guideline regarding best practives for backing up data that can be referenced?

Author: Added notice on backup storage and refer users to their organisations Disaster Recovery (DR) plan and backup policy for further guidelines.

### Manual backup

To manually backup data to current working direcotry

```bash
# >> Using tar zip of the data folder
# This will output the backup file in current directory where it is run
tar -czvf mhira-data-$(date '+%Y-%m-%dT%T').tar.gz ~/.mhira-docker/data
```

To manually backup data to in configured MHIRA backup location

```bash
# >> Using provided backup container
# This will output the backup file in backup path configured in `dotenv` by `DATA_BACKUP_PATH` key

> # Default value for `DATA_BACKUP_PATH` is `~/.mhira-docker/backups`

Reviewer: ```suggestion

```

> docker-compose up  --build -d backup

Reviewer: Not clear why re-building the backup container would produce a manual backup???
Can we discuss this?

Author: This was an error on documenting it. The correct command is

```bash
docker compose exec backup /bin/backup
```
